src/boot/efi/devicetree.c

```diff
@@ -104,6 +113,90 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
                         MAKE_GUID_PTR(EFI_DTB_TABLE), PHYSICAL_ADDRESS_TO_POINTER(state->addr));
 }
 
+static const char* devicetree_get_compatible(const void *dtb) {
+        assert(dtb);
+        const struct fdt_header *dt_header = dtb;
+        if (!IS_ALIGNED64(dt_header) || be32toh(dt_header->Magic) != UINT32_C(0xd00dfeed))
+                return NULL;
```

Comment:
so the C standard says that you a pointer of a specific type that is not aligned to that type is undefined behaviour. Here you cast to `const struct fdt_header*` first, and only then check for alignment. That's hence strictly speaking not correct.

You should check for alignment first (i.e. while the pointer is still a `void*`) and only if that checks out cast it to `const struct fdt_header*`.

Reply:
or in other words:

```c
assert(dtb);

if (!IS_ALIGNED64(dtb))
        return NULL;

const struct fdt_hedaer *dt_header = dtb;

if (be32toh(dt_header->Magic) != UINT32_C(0xd00dfeed))
        return NULL;

…
```

you get the idea

Reply:
```
if (!IS_ALIGNED64(dtb))
        return NULL;

const struct fdt_hedaer *dt_header = ASSERT_PTR(dtb);
...
```